src/benchmarklib/tpch/tpch_query_generator.cpp

```diff
@@ -52,6 +71,330 @@ std::string TPCHQueryGenerator::get_preparation_queries() const {
 }
 
 std::string TPCHQueryGenerator::build_query(const QueryID query_id) {
+  using namespace std::string_literals;
+
+  static std::random_device random_device;
+  // Preferring a fast random engine over one with high-quality randomness
+  static std::minstd_rand random_engine(random_device());
+
+  static std::vector materials{"TIN", "NICKEL", "BRASS", "STEEL", "COPPER"};
+
+  // Random distributions for all strings defined by the TPC-H benchmark
+  static std::uniform_int_distribution<> material_dist{0, static_cast<int>(materials.size() - 1)};
+  static std::uniform_int_distribution<> region_dist{0, regions.count - 1};
+  static std::uniform_int_distribution<> segment_dist{0, c_mseg_set.count - 1};
+  static std::uniform_int_distribution<> nation_dist{0, nations.count - 1};
+  static std::uniform_int_distribution<> type_dist{0, p_types_set.count - 1};
+  static std::uniform_int_distribution<> color_dist{0, colors.count - 1};
+  static std::uniform_int_distribution<> shipmode_dist{0, l_smode_set.count - 1};
+  static std::uniform_int_distribution<> brand_char_dist{1, 5};
+  static std::uniform_int_distribution<> container_dist{0, p_cntr_set.count - 1};
+
+  // This is not nice, but initializing this statically would require external methods and make it harder to
+  // follow in the end. It's not like this list will ever change...
+  static const auto sizes =
+      std::vector{1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+                  26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50};
+  static const auto country_codes =
+      std::vector{10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34};
+
+  std::vector<std::string> parameters;
+
+  switch (query_id) {
+    // Writing `1-1` to make people aware that this is zero-indexed while TPC-H query names are not
+    case 1 - 1: {
+      static std::uniform_int_distribution<> date_diff_dist{60, 120};
```

Comment:
are distributions thread-safe? I am assuming the query generator is called by multiple clients on multiple threads?

Reply:
Good point. The distribution should be fine - to be honest, I am surprised that its `operator()` is not const anyway. The engine is a different thing.

As we are not really concerned about object costs here, I changed them to thread-local variables.